File: ios/Classes/BitlabsPlugin.swift
import BLCustom
import Flutter
import UIKit

public class BitlabsPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "bitlabs", binaryMessenger: registrar.messenger())
        let instance = BitlabsPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        
        BitLabs.shared.setRewardCompletionHandler { reward in
            channel.invokeMethod("onReward", arguments: ["reward": reward])
        }
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "init": initImpl(call, result)
        case "addTag": addTagImpl(call, result)
        case "setTags": setTagsImpl(call, result)
        case "launchOfferWall": launchOfferWallImpl(result)
        case "getSurveys": getSurveysImpl(result)
        case "checkSurveys": checkSurveysImpl(result)
        case "requestTrackingAuthorization": requestTrackingAuthorizationImpl(result)
        default: result(FlutterMethodNotImplemented)
        }
    }
    
    private func initImpl(_ call: FlutterMethodCall, _ result: FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let token = arguments["token"] as? String,
              let uid = arguments["uid"] as? String
        else {
            result(FlutterError(code: "Error", message: "Invalid arguments", details: nil))
            return
        }
        // Assuming 'initialize' is the correct method for BitLabs Initialization
        BitLabs.shared.configure(token: token, uid: uid)
        result("BitLabs initialized")
    }
    
    private func launchOfferWallImpl(_ result: FlutterResult) {
        guard let topViewController = UIApplication.shared.keyWindow?.rootViewController else {
            result(
                FlutterError(
                    code: "Error", message: "Unable to access UIApplication's window", details: nil)
            )
            return
        }
        // Ensure the BitLabs offer wall launch method is correctly used
        BitLabs.shared.launchOfferWall(parent: topViewController)
        result("Offer wall launched")
    }
    
    private func addTagImpl(_ call: FlutterMethodCall, _ result:FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let key = arguments["key"] as? String,
              let value = arguments["value"] as? String
        else {
            result(FlutterError(code: "Error", message: "Invalid arguments", details: nil))
            return
        }
        
        BitLabs.shared.addTag(key: key, value: value)
    }
    
    private func setTagsImpl(_ call: FlutterMethodCall, _ result: FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let tags = arguments["tags"] as? [String: Any]
        else {
            result(FlutterError(code: "Error", message: "Invalid tags argument", details: nil))
            return
        }
        // Assign tags to BitLabs
        BitLabs.shared.setTags(tags)
        result("Tags set")
    }
    
    private func getSurveysImpl(_ result: @escaping FlutterResult) {
        BitLabs.shared.getSurveys { response in
            switch response {
            case .failure(let error):
                result(FlutterError(code: "Error", message: "Get Surveys \(error)", details: nil))
                
            case .success(let surveys):
                if let data = try? JSONEncoder().encode(surveys),
                   let json = try? JSONSerialization.jsonObject(with: data, options: [])
                {
                    result(json)
                }
                
                result(FlutterError(code: "Error", message: "Couldn't Serialize the surveys", details: nil))
            }
        }
    }
    
    private func checkSurveysImpl(_ result: @escaping FlutterResult) {
        BitLabs.shared.checkSurveys { response in
            switch response {
            case .failure(let error):
                result(FlutterError(code: "Error", message: "Check Surveys \(error)", details: nil))
                
            case .success(let surveysExist):
                result(surveysExist)
            }
        }
    }
    
    private func requestTrackingAuthorizationImpl(_ result: @escaping FlutterResult) {
        BitLabs.shared.requestTrackingAuthorization()
        result(true)
    }
}

File: android/src/main/kotlin/ai/bitlabs/bitlabs/BitlabsPlugin.kt
package ai.bitlabs.bitlabs

import android.app.Activity
import ai.bitlabs.sdk.BitLabs
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** BitlabsPlugin */
class BitlabsPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {
  private lateinit var channel : MethodChannel

  private var activity: Activity? = null

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "bitlabs")
    channel.setMethodCallHandler(this)

    BitLabs.setOnRewardListener { reward ->
      channel.invokeMethod("onReward", mapOf("reward" to reward))
    }
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "init" -> initImpl(call, result)
      "setTags" -> setTagsImpl(call, result)
      "launchOfferWall" -> launchOfferWallImpl(result)
      "getSurveys" -> getSurveysImpl(result)
      "checkSurveys" -> checkSurveysImpl(result)
      else -> result.notImplemented()
    }
  }

  private fun initImpl(call: MethodCall, result: Result) = activity?.let {
    val token = call.argument<String>("token") ?: ""
    val uid = call.argument<String>("uid") ?: ""
    BitLabs.init(it, token, uid)
    BitLabs.API.init(token, uid)
    result.success("BitLabs initialised")
  } ?: result.error("Error", "Activity is null", null)

  private fun setTagsImpl(call: MethodCall, result: Result) {
    val tags = call.argument<Map<String, Any>>("tags") ?: run {
      result.error("Error", "Invalid tags argument", null)
      return
    }

    BitLabs.tags = tags.toMutableMap()
    result.success("Tags set")
  }

  private fun launchOfferWallImpl(result: Result) = activity?.let {
    BitLabs.launchOfferWall(it)
    result.success("Offerwall launched")
  } ?: result.error("Error", "Activity is null", null)

  private fun getSurveysImpl(result: Result) = BitLabs.API.getSurveys(
    { surveys ->
      val surveysMap = surveys.map { survey -> survey.toMap() }
      result.success(surveysMap)
    },
    { exception -> result.error("Error", exception.message, null) }
  )

  private fun checkSurveysImpl(result: Result) = BitLabs.API.checkSurveys(
    { surveysExist -> result.success(surveysExist) },
    { exception -> result.error("Error", exception.message, null) }
  )

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  // ActivityAware methods
  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activity = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivity() {
    activity = null
  }
}
